wremnants/CardTool.py

```diff
@@ -355,30 +379,64 @@ def variationName(self, proc, name):
-            perc_sameside = np.count_nonzero(vars_sameside)/hnom.size 
-            if perc_sameside > thresh:
-                logger.warning(f"{perc_sameside:.0%} bins are one sided for syst {name} and process {proc}!")
+            for chan in self.channels:
+                if not self.keepOtherChargeSyst and self.chargeIdDict[chan]["badId"] is not None and self.chargeIdDict[chan]["badId"] in name:
+                    continue
+                if chan in ["plus", "minus"]:
+                    q = self.chargeIdDict[chan]["val"]
+                    hnom = self.getBoostHistByCharge(hnom3D, q)
+                    up = self.getBoostHistByCharge(var_map[name+"Up"], q)
+                    down = self.getBoostHistByCharge(var_map[name+"Down"], q)
+                else:
+                    hnom = hnom3D
+                    up = var_map[name+"Up"]
+                    down = var_map[name+"Down"]
+                nCellsWithoutOverflows = 1
+                for axisLength in hnom.shape:
+                    nCellsWithoutOverflows *= axisLength
+                try:
+                    up_relsign = np.sign(up.values()-hnom.values())
+                except ValueError as e:
+                    logger.error(f"Incompatible shapes between up and down for syst {name}")
+                    raise e
+                down_relsign = np.sign(down.values()-hnom.values())
+                vars_sameside = (up_relsign != 0) & (up_relsign == down_relsign)
+                perc_sameside = np.count_nonzero(vars_sameside)/nCellsWithoutOverflows 
+                if perc_sameside > thresh:
+                    logger.warning(f"Charge {chan.ljust(5)}: {perc_sameside:.0%} bins are one sided for syst {name} and process {proc}!")
+                # check variations are not same as nominal
+                # it evaluates absolute(a - b) <= (atol + rtol * absolute(b))
+                up_nBinsSystSameAsNomi = np.count_nonzero(np.isclose(up.values(), hnom.values(), rtol=1e-07, atol=1e-08))/nCellsWithoutOverflows
+                down_nBinsSystSameAsNomi = np.count_nonzero(np.isclose(down.values(), hnom.values(), rtol=1e-06, atol=1e-08))/nCellsWithoutOverflows
+                if up_nBinsSystSameAsNomi > 0.99 or down_nBinsSystSameAsNomi > 0.99:
+                    logger.warning(f"Channel {chan.ljust(5)}: syst {name} has Up/Down variation with {up_nBinsSystSameAsNomi:.0%}/{down_nBinsSystSameAsNomi:.0%} of bins equal to nominal")
+                    
 
     def writeForProcess(self, h, proc, syst):
-        if syst != self.nominalName and self.systematics[syst]["mirror"]:
-            hnom = self.procDict[proc][self.nominalName]
-            h = hh.extendHistByMirror(h, hnom)
+        decorrelateByCharge = False
+        decorrelateByBin = {}
+        if syst != self.nominalName:
+            systInfo = self.systematics[syst]
+            if systInfo["doActionBeforeMirror"] and systInfo["action"]:
+                h =systInfo["action"](h, **systInfo["actionArgs"])
```

Comment:
I would really rather this happens somewhere else. Why can't it just happen right after the mirror? In principle I expect "writeForProcess" to just write and not apply the transformation

Reply:
Actually that function was already doing some actions, since it calls self.systHists which does the mirror
So it is not true it only writes

Reply:
And it is needed before the mirror because for some systematics there is no systematic yet. This was a particular use case for the fakes where you make a new histogram modifying an existing one (the nominal in that case), so mirroring the nominal to get the alternate doesn't make sense.